test/functional/p2p_blocksonly.py

```diff
@@ -55,8 +55,29 @@ def run_test(self):
         with self.nodes[0].assert_debug_log(['received getdata for: tx {} peer=1'.format(txid)]):
             self.nodes[0].sendrawtransaction(sigtx)
             self.nodes[0].p2p.wait_for_tx(txid)
+            pprint(self.nodes[0].p2p.wait_for_tx(txid))
             assert_equal(self.nodes[0].getmempoolinfo()['size'], 1)
 
+        self.log.info("Restarting node 0 with whitelist permission and blocksonly")
+        self.restart_node(0, ["-persistmempool=0", "-whitelist=127.0.0.1", "-whitelistforcerelay", "-blocksonly"])
+        self.log.info('Check that txs from whitelisted peers are not rejected and relayed to others')
+        self.nodes[0].add_p2p_connection(P2PInterface())
+        peerInfo = self.nodes[0].getpeerinfo()[0]
+        assert_equal(peerInfo['whitelisted'], True)
+        assert_equal(peerInfo['permissions'], ['noban', 'forcerelay', 'relay', 'mempool']) #checkpermission use check permission from p2p_permission
+
+        self.nodes[0].p2p.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+
+        pprint(self.nodes[0].p2p.last_message)
+        pprint(self.nodes[0].p2p.last_message.get('tx'))
+
+        assert_equal(self.nodes[0].p2p.is_connected, True)
+        with self.nodes[0].assert_debug_log(['received getdata for: tx {} peer=0'.format(txid)]):
+            self.nodes[0].sendrawtransaction(sigtx)
```

Comment:
Also, this has been unclear to me:
line 69 suggests to me that it is nodes[0]'s peer who sends out the transaction, which should be received by node[0]. We want to test whether node[0] would accept and relay this tx. This is aligned with purpose of the test.

However, on line 24, 33, and line 76, it seems to be that node[0] is the one who initiates the transaction, zB ` self.nodes[0].sendrawtransaction(sigtx)`. As a relay node (of the peer node[0].p2p), why is node[0] creating, signing, and transaction then?

Reply:
@MarcoFalke  It would be really great if you could clarify this too

Reply:
`self.nodes[0]` is the Bitcoin Core node "in the middle". I suggest not using self.nodes[0].p2p at all and instead assigning the peers proper names. You did that with `first_peer`, `second_peer`.

When you say `self.nodes[0].sendrawtransacttion`, I believe you really meant `first_peer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))`?

Reply:
I think what confuses me is the existing code line 56 uses this construct too: `   self.nodes[0].sendrawtransaction(sigtx)` so I kind of mimicked that line 😂  Does this mean that line 56 should be fixed too?

Reply:
That is a different test "Check that txs from **rpc** are not rejected", here in this line you are testing "Check that the tx from **whitelisted peers** ..."

So those two lines should differ between the test cases